ethcore/src/evm/wasm/runtime.rs

```diff
@@ -0,0 +1,274 @@
+// Copyright 2015-2017 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Wasm evm program runtime intstance
+
+use std::sync::Arc;
+
+use byteorder::{LittleEndian, ByteOrder};
+
+use evm;
+
+use parity_wasm::interpreter;
+use util::{H256, Address};
+
+use super::ptr::{WasmPtr, Error as PtrError};
+use super::call_args::CallArgs;
+
+/// Wasm runtime error
+#[derive(Debug)]
+pub enum Error {
+	/// Storage error
+	Storage,
+	/// Allocator error
+	Allocator,
+	/// Invalid gas state during the call
+	InvalidGasState,
+	/// Memory access violation
+	AccessViolation,
+	/// Interpreter runtime error
+	Interpreter(interpreter::Error),
+}
+
+impl From<interpreter::Error> for Error {
+	fn from(err: interpreter::Error) -> Self {
+		Error::Interpreter(err)
+	}
+}
+
+impl From<PtrError> for Error {
+	fn from(err: PtrError) -> Self {
+		match err {
+			PtrError::AccessViolation => Error::AccessViolation,
+		}
+	}
+}
+
+pub struct Runtime<'a> {
+	gas_counter: u64,
+	gas_limit: u64,
+	dynamic_top: u32,
+	ext: &'a mut evm::Ext,
+	memory: Arc<interpreter::MemoryInstance>,
+}
+
+impl<'a> Runtime<'a> {
+	pub fn with_params<'b>(
+		ext: &'b mut evm::Ext,
+		memory: Arc<interpreter::MemoryInstance>, 
+		stack_space: u32, 
+		gas_limit: u64,
+	) -> Runtime<'b> {
+		Runtime {
+			gas_counter: 0,
+			gas_limit: gas_limit,
+			dynamic_top: stack_space,
+			memory: memory,
+			ext: ext,
+		}
+	}
+
+	pub fn storage_write(&mut self, context: interpreter::CallerContext) 
+		-> Result<Option<interpreter::RuntimeValue>, interpreter::Error>
+	{
+		let mut context = context;
+		let val = self.pop_h256(&mut context)?;
+		let key = self.pop_h256(&mut context)?;
+		trace!(target: "wasm", "storage_write: value {} at @{}", &val, &key);
+
+		// todo: return a runtime error contract can handle or as it is now - general failure?
```

Comment:
I suppose external errors should not be propagated to contracts since they can't handle them in any meaningful way. External error == consensus error here.